### ISSUE TITLE
Option to warn on missing intersphinx config or imports

### ISSUE THREAD

**Issue by DanCardin** [felix-hilden/sphinx-codeautolink#97]
I saw your post on reddit and tried applying it to on of my projects, great plugin!

It's finding a bunch of places where my code examples are either incomplete or missing intersphinx setup. It's the going through and finding all my code examples and trying to click on things that is somewhat tedious.

What would be great is if there was an option to emit warnings where this plugin sees code that it wants to link but can't resolve it. That way I could just run it once and go through and fix all the broken code.

Again, great plugin! Works great!

**Comment by felix-hilden** [comment]
Thank you for the kind words! We do parse all names but just discard the ones that are not found in the object inventory, so this is definitely doable. Not sure if it will be default functionality but I'll have a look!

**Comment by felix-hilden** [comment]
Hmm, I'm thinking that at least some version of this could be default behavior, because it's not really useful if a user finds out about this only after having fixed the issues. We can include a notice on how to disable the warning in the message even, to make it really easy to ignore if it's a known thing.

Having a condensed view of all missing link targets is easy enough to produce and will help users to add the correct intersphinx entries. But that shouldn't be too difficult in the first place given that most libraries only use a handful of other projects in their example code. But making the warning messages appear in each example and link target might be too much spam. Which of these behaviors were you after originally, and what do you think?

Currently I'm playing with having a single message like this:

```
WARNING: Cannot produce links for modules (missing intersphinx entries): 'httpx', 'whatever'
  This warning can be disabled by suppressing warnings of type `codeautolink.locate_object`
  All missing link targets:
    httpx
    httpx.get
    whatever.func
```

**Comment by felix-hilden** [comment]
A quick example of the individual warnings:

```
doc.rst:10: WARNING: Could not find `httpx` when resolving `httpx` on line 2
doc.rst:10: WARNING: Could not find `httpx.get` when resolving `httpx.get` on line 4
```

Maybe we ought to choose one. Having both styles feels like a waste, especially with the detailed view of the summary error. So if both, then let's get rid of the "all missing link targets".

**Comment by mgeier** [comment]
I don't think it would be good to switch this on by default, because there are modules which simply have no `intersphinx` docs, which would lead to superfluous warnings.

I can think of two possible strategies (which are not mutually exclusive) to make this less annoying:

* make this opt-in
* provide a "positive" list of modules which are known to have `intersphinx` docs

If a list of known modules (including their `intersphinx` URL) is available, the warning message could even give a hint how to enable it:

    WARNING: Could not find documentation for `httpx` module, use this to enable:

        intersphinx_mapping = {
            'httpx': ('https://...', None),
        }

**Comment by DanCardin** [comment]
I have no opinion on the default behavior (I want it so i'll turn on the config!), i *can* see how perhaps indviduals' code blocks might intentionally omit imports to make them more compact, or such things and how they might not want this feature. At the same time, it is a warning (not an error).

Re the actual feature, the singular message might be nice, because it enables (as you already did) referencing the config option in the warning itself, which would be a good option if you wanted to enable it by default. But i'm mostly ambivalent about this too. So long as either option shows me all the missing targets, regardless of the reason

Sidebar, it's not purely about intersphinx, plenty of my lacks of links were on my own package were either due to missing imports in the code examples or lack of autodoc'd references to them.

**Comment by felix-hilden** [comment]
Okay to be clear we only try to link via imports. There's no way to guess a module for every unassigned name. So that won't be the behavior 😅 

Thank you both for the comments!

**Comment by DanCardin** [comment]
Oh i didn't expect it to! Not sure which part implied that.

I just wanted to be clear that ideally the mentioned warning wouldn't **just** be missing click targets due to lack of intersphinx, but also missing imports and missing autodoc references

**Comment by felix-hilden** [comment]
What do you mean by missing imports? 😅

**Comment by DanCardin** [comment]
if i write
```
.. code-block::

   from sqlalchemy import create_engine

   def foo():
       url = URL(...)
       return create_engine(url)
```

I would ideally like to see warnings if:
* sqlalchemy isn't set up with intersphinx
* for URL being unknown. sure you can only reasonably produce a link on the URL use if there's an import for you to tack, but i'd love to be able to know that i was missing the `URL` import in the first place
* if sqlalchemy didn't document `create_engine` or `URL`

**Comment by felix-hilden** [comment]
I agree with the first and third options, which are currently displayed in my examples above, but I'm not so keen on the second one.

The reason why I'm somewhat opposed to that is because I don't want to nag about all unassigned names in a code example. They are examples after all. And warning about every unassigned variable seems really excessive. For example, this would produce three warnings, one for each argument:

```python
import httpx
httpx.get(url, method, **params)
```

But if we see an import, it's reasonable to assume that the author wants to know if an entry is missing, either in intersphinx or autodoc, or if the module isn't importable leading to us not finding the proper types.

Does this make sense?

**Comment by DanCardin** [comment]
If it's even technically do-able, i still think it'd be valuable, even if they were separate options. Anecdoally, in all the cases i've encountered (combing through my own examples, trying to click stuff) it's just been a matter of oversight because the docs code isn't being executed/linted, and i'd have preferred to have the imports/symbols so that the code actually executes.

In the above example i think it's entirely reasonable to warn that url/method/params dont exist (if technically possible and an i.e. `url = 'foo'` beforehand would cause that to not error).

The only real counterexample i can think of is in autodoc'd code examples:
```
def foo():
    """
        Examples:
            >>> import foo
            
            some context prose

            >>> foo.bar()
    """
```
doctests interpret this as a single code block that executes with the same locals, but i dont know how that's rendered and whether it'd warn with this theoretical setting

**Comment by felix-hilden** [comment]
I'll do the other warnings first, but I'll think about it some more!

**Comment by felix-hilden** [comment]
I posted a PR! Currently I'm thinking of naming the config `codeautolink_verbose`, but I'm not 100 % on it still. If you have better suggestions or other concerns, I'd be thankful for a review!

I also posted another issue for the unassigned names concern.

**Comment by DanCardin** [comment]
`codeautolink_warn_on_broken_links = True`? (and the other one could be `codeautolink_warn_on_unknown_identifiers`)

If you want it to be more general as verbose fine by me also though

**Comment by felix-hilden** [comment]
If we were to add individual names for each warning, they could be:

- `(codeautolink_warn_on) missing_inventory`
- `... _failed_resolve`
- `... _unassigned_name`

"Broken links" is too ambiguous IMO, and I'm not sure what you meant by "unknown ids". But I'll have to decide if I want to combine some of them.

**Comment by DanCardin** [comment]
these all makes sense to me if you want to individually enumerate them. I agree with all your names (unknown_identifiers (as in variable names) was equivalent to _unassigned_name btw).

**Comment by felix-hilden** [comment]
I think we'll go with the individual names after all. I can't think of another reason as to why they wouldn't be separate other than my personal disliking of making a configuration option for warnings that can also be disabled one by one. But maybe I'll revisit this decision later.

**Comment by DanCardin** [comment]
Upgraded and got:
```
loading pickled environment... done
Traceback (most recent call last):
  File "/Users/danc/work/schireson/pytest-alembic/.venv/lib/python3.8/site-packages/sphinx_codeautolink/extension/__init__.py", line 42, in wrapper
    return func(*args, **kwargs)
  File "/Users/danc/work/schireson/pytest-alembic/.venv/lib/python3.8/site-packages/sphinx_codeautolink/extension/__init__.py", line 87, in build_inited
    self.cache.read()
  File "/Users/danc/work/schireson/pytest-alembic/.venv/lib/python3.8/site-packages/sphinx_codeautolink/extension/cache.py", line 34, in read
    self.transforms[file] = [SourceTransform(**t) for t in transforms]
  File "/Users/danc/work/schireson/pytest-alembic/.venv/lib/python3.8/site-packages/sphinx_codeautolink/extension/cache.py", line 34, in <listcomp>
    self.transforms[file] = [SourceTransform(**t) for t in transforms]
TypeError: __init__() missing 1 required positional argument: 'doc_lineno'
```

Related?

**Comment by felix-hilden** [comment]
Hmm, that's from the cache. Could you try rebuilding from scratch?

**Comment by DanCardin** [comment]
aha, my b. Seems to work, thanks!

**Comment by DanCardin** [comment]
Although, and this may still be user error, I'm seeing:
```
docs/source/experimental_tests.rst:69: WARNING: Inventory missing `sqlalchemy` when resolving `sqlalchemy` on line 1
docs/source/experimental_tests.rst:69: WARNING: Inventory missing `sqlalchemy` when resolving `sqlalchemy` on line 2
docs/source/experimental_tests.rst:69: WARNING: Inventory missing `sqlalchemy.ext.declarative.declarative_base` when resolving `declarative_base` on line 3
WARNING: Cannot locate modules: ..., 'sqlalchemy'
```

for

```
.. code-block:: python

   import sqlalchemy
   from sqlalchemy import Column, types
   from sqlalchemy.ext.declarative import declarative_base
```

with

```
intersphinx_mapping = {
    "sqlalchemy": ("https://docs.sqlalchemy.org/en/14", None),
    ...
```

`Column`, `types`, and `sqlalchemy.ext.declarative` are clickable. Admittedly, in this specific example, `declarative_base` is now documented elsewhere in 14, but it's odd to me that it's emitting the warnings for `sqlalchemy` on both the `import sqlalchemy` as well as the from portion of an import where it actually works on the imported items.

**Comment by felix-hilden** [comment]
This seems odd. Would you mind creating another issue so we can handle this properly?
